watchtower/wtdb/range_index.go

```diff
@@ -0,0 +1,350 @@
+		}
+	}
+
+	return rangeExists, false, rangeStartNum
+}
+
+// lowestHigherRange finds the lowest start value of a range in the set where
+// the start value is still higher than the given value. The first return value
+// is true if such a range exists. The last return value is the start value of
+// the found range.
+func (a *RangeIndex) lowestHigherRange(h uint64) (bool, uint64) {
+	var (
+		rangeExists bool
+		rangeStart  uint64
+	)
+	for height, maxHeight := range a.m {
+		if height <= h {
+			continue
+		}
+
+		if rangeExists && height > rangeStart {
+			continue
+		}
+
+		rangeExists = true
+		rangeStart = height
+
+		// If the new height is smaller than or equal to the max height
+		// of this range, then we can exit early since the height is
+		// covered by this range.
+		if h <= maxHeight {
+			return true, rangeStart
+		}
+	}
+
+	return rangeExists, rangeStart
+}
+
+// MaxHeight returns the highest number covered in the range.
+func (a *RangeIndex) MaxHeight() uint64 {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	return a.maxHeight
+}
+
+// Add adds a single number to the range set.
+func (a *RangeIndex) Add(newHeight uint64) {
+	tx := a.StartTx()
+	tx.Commit(tx.GetChanges(newHeight))
+}
+
+// numRanges returns the number of ranges the set is tracking.
+func (a *RangeIndex) numRanges() int {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	return len(a.m)
+}
+
+// getRangeMax returns the maximum height in the range starting with the given
+// number.
+func (a *RangeIndex) getRangeMax(h uint64) uint64 {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	return a.m[h]
+}
+
+// IndexChanges represents the set of changes that must be performed on a
+// range set in order to add a new number.
+type IndexChanges struct {
+	NewIndexKey   *uint64
+	NewIndexValue *uint64
+	DeleteIndex   *uint64
+}
+
+// RangeIndexTx represents a transaction on the RangeIndex.
+type RangeIndexTx struct {
+	*RangeIndex
+}
+
+// StartTx returns a RangeIndexTx which can be used to perform atomic operations
+// on a RangeIndex. A caller can call GetChanges on the transaction to get the
+// set of changes that need to be performed on the range set in order to add
+// a new number. Commit can then be called to commit then changes or Revert can
+// ge used to cancel the transaction.
+func (a *RangeIndex) StartTx() *RangeIndexTx {
+	a.mu.Lock()
+
+	return &RangeIndexTx{RangeIndex: a}
+}
+
+// Commit can be used to commit a set of IndexChanges to a RangeIndex.
+func (a *RangeIndexTx) Commit(changes *IndexChanges) {
+	defer a.mu.Unlock()
+
+	if changes == nil {
+		return
+	}
+
+	if changes.NewIndexKey != nil && changes.NewIndexValue != nil {
+		a.m[*changes.NewIndexKey] = *changes.NewIndexValue
+
+		if *changes.NewIndexValue > a.maxHeight {
+			a.maxHeight = *changes.NewIndexValue
+		}
+	}
+
+	if changes.DeleteIndex != nil {
+		delete(a.m, *changes.DeleteIndex)
+	}
+}
+
+// Revert can be called to cancel a RangeIndexTx.
+func (a *RangeIndexTx) Revert() {
+	a.mu.Unlock()
```

Comment:
Still diving in, but the mutex usage seems a bit tricky here. 

In this context, would also assume revert would actually wipe/delete the changes?

Reply:
Agreed that the usage is a bit strange - but the idea is that if `Revert` is called then no changes were ever committed to the `RangeIndex`. The caller has to call the above `Commit` method to actually apply any changes

Reply:
gonna change this a bit 👍

Reply:
changed it so that RangeIndex handles the changes (and possible failures) to the persisted DB too 👍